util_linux_test.go

```diff
@@ -0,0 +1,80 @@
+//go:build linux
+
+// Copyright (C) 2024 SUSE LLC. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package securejoin
+
+import (
+	"fmt"
+	"os"
+	"path"
+	"strings"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+)
+
+func requireRoot(t *testing.T) {
+	if os.Geteuid() != 0 {
+		t.Skip("test requires root")
+	}
+}
+
+func withWithoutOpenat2(t *testing.T, testFn func(t *testing.T)) {
+	for _, useOpenat2 := range []bool{true, false} {
+		useOpenat2 := useOpenat2 // copy iterator
+		t.Run(fmt.Sprintf("openat2=%v", useOpenat2), func(t *testing.T) {
+			if useOpenat2 && !hasOpenat2() {
+				t.Skip("no openat2 support")
+			}
+			testForceHasOpenat2 = &useOpenat2
+			defer func() {
+				testForceHasOpenat2 = nil
+			}()
+
+			testFn(t)
+		})
+	}
+}
+
+// Format:
+//
+//	dir <name>
+//	file <name> <?content>
+//	symlink <name> <target>
+func createInTree(t *testing.T, root, spec string) {
+	f := strings.Fields(spec)
+	if len(f) < 2 {
+		t.Fatalf("invalid spec %q", spec)
+	}
+	fullPath := path.Join(root, f[1])
```

Comment:
This makes the tests less interesting. You can't test with a link to "/", for example.

I don't think it is possible to test things outside the root with this, as create test is setting the root to `root` here and we always call `partialOpen()` with the root set by `createTree()`.

So things like a symlink to / or outside the root, we can't test, IIUC.

Reply:
We do have tests with links to `/`. `fullPath` is the path of the symlink, not the target.

Reply:
Oh, unsure if the param order to symlink is reversed to what I was doing and that caused the issue instead :-D

Reply:
Yeah, the spec is `symlink <linkname> <target>` while `symlink(2)` takes the arguments in the opposite order.